x-pack/plugin/old-lucene-versions/src/main/java/org/elasticsearch/xpack/lucene/bwc/codecs/LegacyAdaptingPerFieldPostingsFormat.java

```diff
@@ -0,0 +1,200 @@
+        final SegmentWriteState writeState;
+        final List<Closeable> toClose = new ArrayList<Closeable>();
+
+        FieldsWriter(SegmentWriteState writeState) {
+            this.writeState = writeState;
+        }
+
+        @Override
+        public void write(Fields fields, NormsProducer norms) throws IOException {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void merge(MergeState mergeState, NormsProducer norms) throws IOException {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void close() throws IOException {
+            IOUtils.close(toClose);
+        }
+    }
+
+    private static class FieldsReader extends FieldsProducer {
+
+        private final Map<String, FieldsProducer> fields = new TreeMap<>();
+        private final Map<String, FieldsProducer> formats = new HashMap<>();
+        private final String segment;
+
+        // clone for merge
+        FieldsReader(FieldsReader other) {
+            Map<FieldsProducer, FieldsProducer> oldToNew = new IdentityHashMap<>();
+            // First clone all formats
+            for (Map.Entry<String, FieldsProducer> ent : other.formats.entrySet()) {
+                FieldsProducer values = ent.getValue().getMergeInstance();
+                formats.put(ent.getKey(), values);
+                oldToNew.put(ent.getValue(), values);
+            }
+
+            // Then rebuild fields:
+            for (Map.Entry<String, FieldsProducer> ent : other.fields.entrySet()) {
+                FieldsProducer producer = oldToNew.get(ent.getValue());
+                assert producer != null;
+                fields.put(ent.getKey(), producer);
+            }
+
+            segment = other.segment;
+        }
+
+        FieldsReader(final SegmentReadState readState, LegacyAdaptingPerFieldPostingsFormat legacyAdaptingPerFieldPostingsFormat)
+            throws IOException {
+
+            // Read _X.per and init each format:
+            boolean success = false;
+            try {
+                // Read field name -> format name
+                for (FieldInfo fi : readState.fieldInfos) {
+                    if (fi.getIndexOptions() != IndexOptions.NONE) {
+                        final String fieldName = fi.name;
+                        final String formatName = fi.getAttribute(PER_FIELD_FORMAT_KEY);
+                        if (formatName != null) {
+                            // null formatName means the field is in fieldInfos, but has no postings!
+                            final String suffix = fi.getAttribute(PER_FIELD_SUFFIX_KEY);
+                            if (suffix == null) {
+                                throw new IllegalStateException("missing attribute: " + PER_FIELD_SUFFIX_KEY + " for field: " + fieldName);
+                            }
+                            PostingsFormat format = legacyAdaptingPerFieldPostingsFormat.getPostingsFormat(formatName);
```

Comment:
I think we introduced this adapter because of line 125? I tried to remove this class by converting the `PerFieldPostingsFormat.format` attribute from `Lucene50` to `BWCLucene50` in FieldInfos, but it didn't work because the line 126. I think it's fine to have this class.

Reply:
Another motivation for this class is that it keeps the information in these older indices as much "as-is" as possible, not requiring a conversion step before reading. Furthermore, future versions won't have to be aware that that conversion step happened in some prior versions of archive, if we were to remove it e.g. in the future. The nice bit is that this conversion can go away in ES 9 as we won't have a conflicting Lucene50PostingsFormat shipped anymore as part of bwc jars.